x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/grouping/Categorize.java

```diff
@@ -156,4 +176,175 @@ public Expression field() {
     public String toString() {
         return "Categorize{field=" + field + "}";
     }
+
+    public GroupingKey.Supplier groupingKey(Function<Expression, ExpressionEvaluator.Factory> toEvaluator) {
+        return mode -> new GroupingKeyFactory(source(), toEvaluator.apply(field), mode);
```

Comment:
```
... | STATS ... BY SOME_FUNCTION(CATEGORIZE(field))
```
is very interesting. _For now, I think we should just consider this invalid_, but there's options to make this work:
1. `CATEGORIZE` actually is a pipeline breaker, i.e. it needs to be finalized on the coordinator. We could declare it its own, standalone physical execution plan node, and the execution plans could be something like `categorize (initial) -> exchange (send to coord.) -> categorize (final) -> eval (some_function) -> stats (on the coordinator)`. In cases without a function on top of the categorize, the optimizer could fuse the `stats` and the `categorize` into one combined node.
2. We could use a 2-phase approach, where we determine the full categorizer in the first phase, so that `categorize` can be executed like a regular function in the second phase.

```
... | STATS ... BY CATEGORIZE(SUBSTRING(field, 10))
... | EVAL s = SUBSTRING(field, 10) | STATS ... BY CATEGORIZE(s)
```
For this equivalence, our optimizer already has a mechanism to turn the former into the latter; we just need to teach it about CATEGORIZE.